include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -32,6 +32,117 @@ namespace LinearAlgebra
+          src.trilinos_vector(), dst.trilinos_vector(), mode, alpha, beta);
+      }
+
+
+
+      template <typename Number, typename MemorySpace>
+      void
+      apply(const SparseMatrix<Number, MemorySpace> &M,
+            const dealii::Vector<Number>            &src,
+            dealii::Vector<Number>                  &dst,
+            Teuchos::ETransp                         mode = Teuchos::NO_TRANS,
+            Number alpha = Teuchos::ScalarTraits<Number>::one(),
+            Number beta  = Teuchos::ScalarTraits<Number>::zero())
+      {
+        Assert(&src != &dst,
+               SparseMatrix<double>::ExcSourceEqualsDestination());
+        Assert(M.trilinos_matrix().isFillComplete(),
+               SparseMatrix<double>::ExcMatrixNotCompressed());
+
+#  ifdef DEBUG
+        // TODO: If the host spaces do not match, we can copy the memory
+        //       from one device to an other.
+        bool memory_space_matches =
+          std::is_same_v<typename MemorySpace::kokkos_space,
+                         typename Kokkos::HostSpace>;
+        Assert(memory_space_matches, ExcInternalError());
+#  endif
+
+        // get the size of the input vectors:
+        const size_type dst_local_size = dst.end() - dst.begin();
+        const size_type src_local_size = src.end() - dst.begin();
+
+        // For the dst vector:
+        Kokkos::View<
+          Number **,
+          Kokkos::LayoutLeft,
+          Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
+                         typename MemorySpace::kokkos_space>>
+          kokkos_view_dst(dst.begin(), dst_local_size, 1);
+
+        // get a Kokkos::DualView
+        auto mirror_view_dst = Kokkos::create_mirror_view(kokkos_view_dst);
+        typename SparseMatrix<Number, MemorySpace>::VectorType::dual_view_type
+          kokkos_dual_view_dst(kokkos_view_dst, mirror_view_dst);
+
+        // create the Tpetra::Vector
+        typename SparseMatrix<Number, MemorySpace>::VectorType tpetra_dst(
+          M.trilinos_matrix().getRangeMap(), kokkos_dual_view_dst);
+
+        // For the src vector:
+        // create a Kokkos::View from the src vector
+        Kokkos::View<
+          Number **,
+          Kokkos::LayoutLeft,
+          Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
+                         typename MemorySpace::kokkos_space>>
+          kokkos_view_src(const_cast<Number *>(src.begin()), src_local_size, 1);
+
+        // get a Kokkos::DualView
+        auto mirror_view_src = Kokkos::create_mirror_view(kokkos_view_src);
+        typename SparseMatrix<Number, MemorySpace>::VectorType::dual_view_type
+          kokkos_dual_view_src(kokkos_view_src, mirror_view_src);
+
+        // create the Tpetra::Vector
+        typename SparseMatrix<Number, MemorySpace>::VectorType tpetra_src(
+          M.trilinos_matrix().getDomainMap(), kokkos_dual_view_src);
+
+        M.trilinos_matrix().apply(tpetra_src, tpetra_dst, mode, alpha, beta);
```

Comment:
Does this also `sync` the `DualView`?

Reply:
I just assumed it does. I just verified it passed the test:
```
trilinos_tpetra/sparse_matrix_vector_08.cc
```
Since it passed that test, I assumed it was okay as it was.